File: Exam_Practice/BinarySearchTree.cpp
#include<iostream>
using namespace std;
class Node
{
    private :
    int data;
    Node* left;
    Node* right;
    public :
        Node(int data){
            this->data=data;
            this->left=NULL;
            this->right=NULL;
        }
        Node(){
            this->data=0;
            this->left=NULL;
            this->right=NULL;
        }
        void setData(int data){
            this->data=data;
        }       
        int getData(){
            return data;
        }
        Node* getLeft(){
            return left;
        }
        void setLeft(Node* left)
        {
            this->left=left;
        }
        Node* getRight(){
            return this->right;
        }
        void setRight(Node* right){
            this->right=right;
        }
};
class Queue{
    private :
    int front;
    int rear;
    int size;
    int* store;
    public:
    Queue(int size){
        this->size=size;
        front=rear=-1;
        store=new int[size];
    }
    Queue(){
        size=20;
        front=rear=-1;
        store=new int[size];
    }
    void Enqueue(int data){
            store[++rear]=data;    
    }
    ~Queue(){delete[] store;}
    int Dequeue(){
            int data = store[++front];
            return data;
    }
    
    bool isEmpty(){
        if(front == rear){
            front = rear=-1;
            return true;
        }
        return false;
    }
    bool isFull(){
        if(rear == (size -1) ){
            return true;
        }
        return false;
    }
};
class  Tree
{
private:
    Node* root;
public:
    Tree()
    {
        root = NULL;
    }
    Node* getRoot(){return root;}
    void insertData(int data)
    {
       
        Node* node = new Node(data);
        Node* temp=root;
        Node* prev=root;
        if(root==NULL)
        {// cout<<"inside"<<endl;
            root=node;
        }
        else
        {
            while (temp!=NULL)
            {
               // cout<<"While loop "<<endl;
                if(temp->getData() >= data){
                    prev=temp;
                   temp = temp->getLeft();
                }
                else if(temp->getData() <= data)
                {
                    prev=temp;
                    temp=temp->getRight();
                }
            } //while loop end
            if(prev->getData() <= data){
                
               // cout<<"right"<<endl;
                prev->setRight(node);
            }
            else
            {
               // cout<<"left"<<endl;
                prev->setLeft(node);
            }   
        } //else
    } //function

    ~Tree()
    {
        root=NULL;
    }
    void Print(){
        ShowInorder(root);
    }
    void ShowInorder(Node* temp){
        if (temp!=NULL) 
        {
 
            if(temp->getLeft()!=NULL){
                ShowInorder(temp->getLeft());
            }
            cout << temp->getData() <<" ";
            if(temp->getRight()!=NULL){
                ShowInorder(temp->getRight());
            }

        }      
    }
    void LevelOrder(){
       
        Queue q(10);
        
        Node* temp=root;
        if(temp!=NULL){ 
            q.Enqueue(temp->getData());
            while (!q.isEmpty())
            {
                int data = q.Dequeue();
                cout<<data<<endl;
              if(temp->getLeft()!=NULL){
                q.Enqueue(temp->getLeft()->getData());
                temp=temp->getLeft();
              }
                if(temp->getRight()!=NULL){
                q.Enqueue(temp->getRight()->getData());
                temp=temp->getRight();
                }
            }
            
        }

    }
    void Search(int data){
        //Find(root,data);
        FindPath(root,data);
    }
    void Find (Node* root , int data){
       if(root==NULL){
           cout<<"Not found"<<endl;
       }else
       {
            if(data < root->getData()){
                return Find(root->getLeft(),data);
            }
            if(data > root->getData()){
                return Find(root->getRight(),data);
            }
            else
            {
                cout<<"Found"<<endl;
            }
            
       }
       
        
    }
      void FindPath (Node* root , int data){
       if(root==NULL){
           cout<<"Not found"<<endl;
       }else
       {
           cout<<"Root "<<endl;

            if(data < root->getData()){
                Find(root->getLeft(),data);
           cout<<"L "<<endl;
                
            }
            if(data > root->getData()){
                Find(root->getRight(),data);
           cout<<"R "<<endl;
            }
            else
            {
                cout<<"Found"<<endl;
            }   
       }
    }
Node FindMax(Node* root){
   Node temp;
   Node RTemp,LTemp;
   int max = root->getData();
   if(root==NULL){
       return INT_MIN ;
   }
   else if(root->getLeft()!=NULL){
       LTemp = FindMax(root->getLeft());
   }
   else if(root->getRight()!=NULL){
       RTemp= FindMax(root->getRight());
   }
   if(LTemp.getData() > RTemp.getData()){
         temp= LTemp;
   }
   else
   {
       temp=RTemp;
   }
   return temp;
}
    Node* DeleteNode(Node* root,int data){
        Node temp;
        if(root == NULL){
            cout<<"No Data Exist"<<endl;
        }
        else if(data<root->getData()){
            root->setLeft(DeleteNode(root->getLeft(),data));
        }
        else if(data > root->getData()){
            root->setRight(DeleteNode(root->getRight(),data));
        }
        else if(root->getLeft()!=NULL && root->getRight()!=NULL){
            temp= FindMax(root->getLeft());
            root->setData(temp.getData());
            root->setLeft(DeleteNode(root->getLeft(),root->getData()));
        }
        else{
            Node* temp =root;
            if(root->getLeft() == NULL){
                root->getRight();
            }
            if(root->getRight() == NULL){
                root = root->getLeft();
            }
            delete temp;
        }
        return new Node();
    }
};



int main()
{
   
    Tree t ;
    // t.insertData(34);
    // t.insertData(23);
    // t.insertData(56);
    // t.insertData(6);
    // t.insertData(80);
    t.insertData(4);
    t.insertData(2);
    t.insertData(8);
    t.insertData(1);
    t.insertData(5);
    t.insertData(9);
    t.insertData(7);
    t.insertData(6);
    

    t.DeleteNode(t.getRoot(),5);
     t.Print();
   // t.LevelOrder();

    //t.Search(12);
    
    
    
    
     
     

     
     
    return 0;
}

File: Interview_practice/src/TreesInterviewQuestions/BinarySearchTree/BinarySearchTree.java
package TreesInterviewQuestions.BinarySearchTree;

import java.util.LinkedList;
import java.util.Queue;

public class BinarySearchTree {
    BinaryNode root;
    int depth;

    public BinarySearchTree() {
        root = null;
        depth = 0;
    }

    private BinaryNode insertNode(BinaryNode currentNode, int value) {
        if (currentNode == null) {
            BinaryNode newBinaryNode = new BinaryNode(value);
            return newBinaryNode;
        } else if (currentNode.value >= value) {
            currentNode.left = insertNode(currentNode.left,value);
            return currentNode;
        }else{
            currentNode.right = insertNode(currentNode.right,value);
            return currentNode;
        }
    }
    
    public void insert(int value){
       root = insertNode(root,value);
    }

    public void preOrder(BinaryNode root){
        if(root == null) return;
        System.out.print(root.value + " " );
        preOrder(root.left);
        preOrder(root.right);
    }

    public void postOrder(BinaryNode root){
        if(root == null) return;
        postOrder(root.left);
        postOrder(root.right);
        System.out.print(root.value + " " );
    }

    public void inOrder(BinaryNode root){
        if(root == null) return;
        inOrder(root.left);
        System.out.print(root.value + " " );
        inOrder(root.right);
    }

    public void levelOrder(BinaryNode root){
        Queue<BinaryNode> queue = new LinkedList<>();
        queue.add(root);
        while(!queue.isEmpty()){
            BinaryNode node= queue.remove();
            System.out.print(node.value + " ");
            if(node.left != null){
                queue.add(node.left);
            }
            if(node.right != null){
                queue.add(node.right);
            }
        }
    }

    public BinaryNode search(BinaryNode root,int value){
        if(root.value == value) return root;
        if(root.value>=value) {
            System.out.print("left->");
           root = search(root.left,value);

        }else{
            System.out.print("right->");
           root = search(root.right,value);

        }
        return root;
    }

    public BinaryNode minimumNode(BinaryNode node){
        if(node.left == null){
            return node;
        }else{
            return minimumNode(node.left);
        }
    }

    public BinaryNode deleteNode(BinaryNode root , int value){
        if(root == null){
            System.out.println("value not found Tree is Empty");
            return null;
        }
        if(root.value < value){
            root.right = deleteNode(root.right,value);
        }else if(root.value > value){
           root.left = deleteNode(root.left,value);
        }else{
            if(root.left != null && root.right != null){
                BinaryNode temp = root;
                BinaryNode rightSuccessor = minimumNode(temp.right);
                root.value = rightSuccessor.value;
                root.right = deleteNode(root.right,rightSuccessor.value);
            }else if(root.left != null && root.right == null){
                root = root.left;
            }else if(root.left == null && root.right != null){
                root = root.right;
            }else{
                root = null;
            }
        }
        return root;
    }

    public static void main(){
        BinarySearchTree b = new BinarySearchTree();
        b.insert(70);
        b.insert(50);
        b.insert(90);
        b.insert(30);
        b.insert(60);
        b.insert(80);
        b.insert(100);
        b.insert(20);
        b.insert(40);
        b.preOrder(b.root);
        System.out.println();
        b.inOrder(b.root);
        System.out.println();
        b.postOrder(b.root);
        System.out.println();
        b.levelOrder(b.root);
        System.out.println();
        System.out.println(b.search(b.root,40).value);

        System.out.println(b.deleteNode(b.root,80));
        b.inOrder(b.root);
    }
}
